ocaml/xapi/xapi_periodic_scheduler.ml

```diff
@@ -30,9 +30,37 @@ let delay = Delay.make ()
 let (queue : (t Ipq.t)) = Ipq.create 50
 let lock = Mutex.create ()
 
+module Clock : sig
+  (** Oclock represents time in nanoseconds in a 64-bit integer. We
+   * can't make this abstract because module Ipq relies
+   * on it (for now). Note that the monotonic clock is only good for
+   * relative time measurements and the values returned by two different
+   * clocks like Oclock.monotonic and Oclock.realtime are unrelated. The
+   * underlying mechanism is clock_gettime(2).
+   *)
+  type t = Int64.t
+  val gettime:  unit -> t
+  val add:      t -> float -> t
+  val sub:      t -> t -> t
+  val to_sec:   t -> float
+end = struct
+  type t             = Int64.t
+  let clock          = Oclock.monotonic
+  let nano           = 1_000_000_000L (* 1E9 *)
+  let nano'          = Int64.to_float nano
+  let gettime ()     = Oclock.gettime clock
+  let add time delay = Int64.(add time (mul (of_float delay) nano))
+  let sub t1 t2      = Int64.(sub t1 t2)
+  let to_sec delay   = (Int64.to_float delay) /.  nano'
+end
+let (++) = Clock.add
```

Comment:
I wonder if `(++)` might be nicer inside the the `Clock` module, so that client code can do things like
`time=Clock.(gettime () ++ start)`

Reply:
Also I wondered about a function in Clock like
```OCaml
let now_plus delay =
  add (gettime ()) delay
```
for convenience in use-cases like the example above, but it's probably not worth it given that gettime and addition are each still needed separately.

Reply:
`++` can already be used as in `Clock.( ... ++ ...)` but by putting it inside the module it would be enforced.

Reply:
Sorry, yes, I didn't write quite what I meant. I didn't mean to suggest it wasn't already possible.